Source/Diagnostics/ReducedDiags/ColliderRelevant.cpp

```diff
@@ -0,0 +1,559 @@
+    for (int i_s = 0; i_s < 2; ++i_s)
+    {
+        // get WarpXParticleContainer class object
+        WarpXParticleContainer& myspc = mypc.GetParticleContainerFromName(m_beam_name[i_s]);
+        // get charge
+        amrex::ParticleReal const q = myspc.getCharge();
+
+        using PType = typename WarpXParticleContainer::SuperParticleType;
+
+        num_dens[i_s] = myspc.GetChargeDensity(0);
+        num_dens[i_s]->mult(1./q);
+
+        // wtot
+        amrex::Real wtot = ReduceSum( myspc,
+            [=] AMREX_GPU_HOST_DEVICE (const PType& p)
+            {
+                return p.rdata(PIdx::w);
+            });
+        amrex::ParallelDescriptor::ReduceRealSum(wtot, amrex::ParallelDescriptor::IOProcessorNumber());
+
+#if defined(WARPX_DIM_XZ)
+        // thetax_min, thetax_max
+        amrex::Real thetax_min = 0.0_rt;
+        amrex::Real thetax_max = 0.0_rt;
+
+        amrex::ReduceOps<ReduceOpMin, ReduceOpMax> reduce_ops_minmax;
+        auto r_minmax = amrex::ParticleReduce<amrex::ReduceData<Real, Real>>(
+            myspc,
+            [=] AMREX_GPU_DEVICE(const PType& p) noexcept -> amrex::GpuTuple<Real, Real>
+            {
+                const amrex::Real ux = p.rdata(PIdx::ux);
+                const amrex::Real uz = p.rdata(PIdx::uz);
+                const amrex::Real thetax = std::atan2(ux, uz);
+                return {thetax, thetax};
+            },
+            reduce_ops_minmax);
+        thetax_min = amrex::get<0>(r_minmax);
+        thetax_max = amrex::get<1>(r_minmax);
+        amrex::ParallelDescriptor::ReduceRealMin(thetax_min, amrex::ParallelDescriptor::IOProcessorNumber());
+        amrex::ParallelDescriptor::ReduceRealMax(thetax_max, amrex::ParallelDescriptor::IOProcessorNumber());
+
+        // x_ave, thetax_ave
+        amrex::Real x_ave = 0.0_rt;
+        amrex::Real thetax_ave = 0.0_rt;
+
+        amrex::ReduceOps<ReduceOpSum, ReduceOpSum> reduce_ops_ave;
+        auto r_ave = amrex::ParticleReduce<amrex::ReduceData<Real, Real>>(
+            myspc,
+            [=] AMREX_GPU_DEVICE(const PType& p) noexcept -> amrex::GpuTuple<Real, Real>
+            {
+                const amrex::Real w  = p.rdata(PIdx::w);
+                const amrex::Real x = p.pos(0);
+                const amrex::Real ux = p.rdata(PIdx::ux);
+                const amrex::Real uz = p.rdata(PIdx::uz);
+                const amrex::Real thetax = std::atan2(ux, uz);
+                return {w*x, w*thetax};
+            },
+            reduce_ops_ave);
+        x_ave = amrex::get<0>(r_ave);
+        thetax_ave = amrex::get<1>(r_ave);
+        amrex::ParallelDescriptor::ReduceRealSum({x_ave, thetax_ave}, amrex::ParallelDescriptor::IOProcessorNumber());
+        x_ave = x_ave / wtot;
+        thetax_ave = thetax_ave / wtot;
+
+        // x_std, y_std, thetax_std, thetay_std
+        amrex::Real x_std = 0.0_rt;
+        amrex::Real thetax_std = 0.0_rt;
+
+        amrex::ReduceOps<ReduceOpSum, ReduceOpSum> reduce_ops_std;
```

Comment:
We might be able to avoid a few of the repeated reduces over all particles if we reformulate some of the statistical quantities needed. I sent you a PDF with the concept.

Reply:
If I understand correctly, the idea is to compute as many quantities as possible with a single pass over all the particles. 

In this case, the quantities that the diagnostic calculates are simple averages and standard deviations over all the particles of the relevant species (e.g. there are not cross-correlations). 

To avoid doing multiple passes, I understand that the idea would be to use the formula for the variance:
`Var(X) = <X^2> - <X>^2` 
so to get in a single pass `<X^2>` and `<X>^2` (`<*>` is the weighted average) and then the variance without having to loop again through the particles. 

Is this what you are suggesting?

Isn't this calculation prone to numerical cancellation? 

On a side note, should I also compute the minimum and maximum angles together with the averages?

Reply:
Yes, it is generally better to have few kernel launches. Note that the same is true for the parallel operations. Those should also be combined as much as possible. See for example PR #4050.

That's a good question regarding numerical cancellation. I don't know if doing the calculation that way causes problems, but this is the way that Warp had done it and I don't remember it ever being an issue.